Support 15 servos on DroneCAN

henrykotze opened the pull request:
Currently only 8 servos can be command simultaneously over DroneCAN, but `uavcan::equipment::actuator::ArrayCommand` takes up to 15 Servo commands (`uavcan::equipment::actuator::Command`) 

This changes bring the number of servos which can be commanded simultaneously to 15.

## Testing

Using a Hitec Servo and setting its ID to 15, and by using the actuator tab on QGC it was following the slider commands as expected.

Commit:
Support 15 servos on CAN

Commit:
Support up to 15 ESCs

- The DroneCAN ESC message supports up to 20 ESC'. Push the number of
ESC's to this limit.

henrykotze commented:
Increased the number of ESC's being  controlled to the max number supported on the DroneCAN message. I have not tested the ESC implementation yet.

danielhonies commented:
For the ESCs QGroundcontrol shows this: 
![image](https://user-images.githubusercontent.com/10270416/233985038-99086e43-57c9-4579-8125-16ed6dc81d63.png)

danielhonies commented:
For the ESCs Qgroundcontrol shows this: 
![image](https://user-images.githubusercontent.com/10270416/233985226-87e63314-7444-4645-898b-32ee7df2617a.png)

danielhonies commented:
Up to ESC index 15, I can confirm it works with the Qgroundcontrol actuator slider test.

henrykotze commented:
@danielhonies Thanks for testing! 
I extended the ESC range up to 20. Do you know why it gave those param issues for 16-20?

danielhonies commented:
I haven't found out yet.

Commit:
Increased MAX_ACTUATORS to 20 in mixer

- This changes PWM_OUTPUT_MAX_CHANNELS to 20, because the MAX_ACTUATORS
in the mixer module is a constexpr from this value

henrykotze commented:
QGC will now show up  to 20 motors and still only 8 for servos. I had to update the module.yaml inside the control_allocator to increase the number of motors being selected in QGC (`CA_ROTOR_COUNT` resulted in this. Refer to **A** in the picture) . To be honest, I dont think its  necessary to increase the number of motors up to this limit, but perhaps its good, because it shows PX4 is capable of doing it.

However the parameters to define the motor position from 12-20 is not being generated currently and thus will show parameters not available (refer to **B**) Thus CA_ROTOR**x**_* is only being generated up to x = 12

The parameters not being available was due to MAX_ACTUATOR inside the mixer module being limited to 16 (refering to your picture being posted). This 16 is based on the number of PWM outputs. I increased it to 20 to resolve QGC not finding the parameters. (refer to **C**) 

It seems there are a few side effects caused by increasing the limit, so we might need think what would be the best approach here. 

Will try and see if I can find out why its not generating those parameters, but this will probably only be next week. 

![image](https://user-images.githubusercontent.com/24815968/234382775-a943c2de-1b02-4ec8-b39b-62be447437c0.png)

mrpollo commented:
hey @dagar do you know where the limitation is coming from here? @pavel-kirienko is this a DroneCan limitation?

pavel-kirienko commented:
DroneCAN supports at most 15 servos and at most 20 motors. These limitations are hard to circumvent. More on this here: https://forum.opencyphal.org/t/cyphal-vs-dronecan/1814

junwoo091400 commented:
Response from beat on [discord](https://discord.com/channels/1022170275984457759/1100809711131041882/1100996978940325940):

> Yes uorb message size and flash due to more parameters. We can increase it but need to check how these numbers are affected.

Commit:
Revert number of ESC to 16

henrykotze commented:
Starting from CA_ROTOR12 which reflect to MOTOR 13 in QGC, shows up as param not available (as seen from the screenshot above). I then check whether CA_ROTOR12 and above is being generated,  and can confirm it does exist in the `parameters.xml` file, but also in the autogenerated enum class `px4:params`

I than tried to see if I can find the parameter via the mavlink console, and stumble upon this: 
![image](https://github.com/PX4/PX4-Autopilot/assets/24815968/f9ea58ce-a869-4c28-b9b5-44ceaed8fe11)

It seems like parameters above 256 is unable to fetch, since param index 256  reflects to CA_ROTOR12_AX
This would correspond to the param_count function return type is of type unsigned, and not uint16_t. https://github.com/PX4/PX4-Autopilot/blob/8222cbca2ce5ffa43e2c542fc125dea4d69356ac/src/lib/parameters/param.h#L114

I have changed the returned type to uint16_t, but remain with the same problem as above. 

Still investigating, any comments are welcome. :)

henrykotze commented:
After setting the parameters in the mavlink console, and then existing QGC, and then opening it again, I was able to get QGC to show the parameters which I set, which was CA_ROTOR12 and CA_ROTOR13. 

![image](https://github.com/PX4/PX4-Autopilot/assets/24815968/d8bd22ab-6a20-4f62-bdac-2808ef644ba0)

I will look further tomorrow.

Commit:
Testing - setting rotors params to display in qgc

henrykotze commented:
The reason why QGC showed the (Param not available) is because if the parameters is not set, it is not sent to QGC. 

For testing purposes, I set the remaining rotor parameters to all zero in the Generic Dodecarotor cox geometry. When choosing this airframe and going to the actuators tab in QGC, and now selecting rotors to be 16, will results in them showing up.

![image](https://github.com/PX4/PX4-Autopilot/assets/24815968/4406adff-1784-4a9b-8b76-7e344b0fecdf)

LeTo37 commented:
Hi all,

The inclusion of more DroneCAN servos has become increasingly useful for us personally and so I was just wondering whether there had been any resolution on this discussion: https://discuss.px4.io/t/px4-community-q-a-april-26-2023/31839/2
And any particular reason why this PR has not yet been reviewed?

Thanks for the awesome work @henrykotze @danielhonies !

mrpollo commented:
@dakejahl, if you are interested, and can point me to where to purchase some, I can send some your way

dakejahl commented:
> @dakejahl, if you are interested, and can point me to where to purchase some, I can send some your way

I'm sorta interested, it might come up with our customers down the road and it would be nice to point them in a direction. That being said I'm not sure what is commonly/currently being used. At the very least they should be DroneCAN + NDAA. I could ask in our Discord about it, but it's not a high priority for us at the moment.

henrykotze commented:
@mrpollo We bought a few from Hitec, Zeus,  Himark and MKS servos.
This was a few years ago, but the most plug and play solution at that stage were the MKS servos, with being able to set the parameters the most easily over the Dronecan gui app.

Hitec provides their own GUI to adjust these parameters, but we have seen there were conflict with the Servo ids when you wanted to control more than 1.

We bought Himark servos, but they used a different/non standard DSDL message to control their servos which made integration cumbersome when you have different manufacturing servo's on your drone. (For different load bearing situations)

Cant say much about Zeus from which i can remember. 

All the above is from testing a few years ago, so this is hopefully outdated.
